<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Promise:非同期処理</title>
</head>
<body>
<script>

/*2.6. Promise#finally

  ECMAScript2018からpromise chainの最後に処理を実行する 
  Promise#finally メソッド追加されました。

  Promise#finallyメソッドは成功時、失敗時どちらの場合でも
  呼び出すコールバック関数を登録できます。 
  
  try…​catch…​finally構文のfinally節と同様の役割をもつメソッドです。

  次のコードのように、 Promise#finallyメソッドで登録したコールバック関数は、
  promiseオブジェクトがresolve(成功) / reject(失敗)どちらの場合でも呼ばれます。

  finally メソッドのコールバック関数は引数を受け取らず、
  どのような値を返してもpromise chainには影響を与えません。 
    
  また、 finallyメソッドは新しいpromiseオブジェクトを返し、
  新しいpromiseオブジェクトは呼び出し元のpromiseオブジェクトの状態を
  そのまま引き継ぎます。
  
  // finallyとpromise chain

  function onFinally() {
      // 成功、失敗どちらでも実行したい処理
  }
  
  // `Promise#finally` は新しいpromiseオブジェクトを返す
  Promise.resolve(42)

  .finally(onFinally)
    
  .then((value) => {
        // 呼び出し元のpromiseオブジェクトの状態をそのまま引き継ぐ
        // 呼び出し元のpromiseオブジェクトは `42` で resolveされている
        console.log(value); // 42
  });
  
  /*実行結果:Google Chrome-console

    42  
  */
 
  /*Promise#finallyメソッドと同等の表現をPromise#thenメソッドで書くと
    次のように書けます。
  
  // finallyをthenで表現

  function onFinally() {
      // 成功、失敗どちらでも実行したい処理
  }

  // Promise#finally(onFinally)と同等の表現
  promise.then((result) => {

      onFinally();

      return result;

  }, (error) => {

      onFinally();

      return Promise.reject(error);

  });

  Promise#finallyメソッドを使うことで、
  promise chainで必ず実行したい処理を簡単に書けるようになっています。

  次のコードでは、リソースを取得中かどうかを判定するためのフラグを
  isLoadingという変数で管理しています。
  
  Promise#finally メソッドを使い成功失敗どちらにもかかわらず、
  取得が終わったらisLoadingはfalseにしています。

  finallyのユースケース*/

  // リソースを取得中かどうかのフラグ
  let isLoading = false;

  function fetchResource(URL) {
      
    // リソース取得中フラグをONに
    isLoading = true;
      
    return new Promise((resolve, reject) => {
          
      const req = new XMLHttpRequest();
          
      req.open("GET", URL, true);
          
      req.onload = () => {
              
        if (200 <= req.status && req.status < 300) {
                  
          resolve(req.responseText);
              
        } else {
                  
          reject(new Error(req.statusText));
              
        }
          
      };
          
      req.onerror = () => {
              
        reject(new Error(req.statusText));
          
      };
          
      req.send();
      
    }).finally(() => {
          
      // リソース取得に成功/失敗どちらの場合も取得中フラグをOFFに
      isLoading = false;
      
    });
  }

  console.log("リソースロード開始", isLoading);

  fetchResource("https://httpbin.org/get").then((value) => {

      console.log("リソース取得に成功", isLoading);

      console.log(value);

  }).catch((error) => {

      console.log("リソース取得に失敗", isLoading);

      console.error(error);

  });

  console.log("リソースロード中", isLoading);

  /*thenとcatchメソッドでも実現できますが、 
    Promise#finallyメソッドを使うことでisLoading の代入を一箇所にまとめられます。
  
  /*実行結果:Google Chrome-console
    リソースロード開始 false
    リソースロード中 true
    リソース取得に成功 false
    {
      "args": {}, 
      "headers": {
        "Accept": "/*//*", 
        "Accept-Encoding": "gzip, deflate, br", 
        "Accept-Language": "ja,en-US;q=0.9,en;q=0.8", 
        "Dnt": "1", 
        "Host": "httpbin.org", 
        "Origin": "http://127.0.0.1:5500", 
        "Referer": "http://127.0.0.1:5500/", 
        "Sec-Ch-Ua": "\"Google Chrome\";v=\"123\", \"Not:A-Brand\";v=\"8\", \"Chromium\";v=\"123\"", 
        "Sec-Ch-Ua-Mobile": "?0", 
        "Sec-Ch-Ua-Platform": "\"Windows\"", 
        "Sec-Fetch-Dest": "empty", 
        "Sec-Fetch-Mode": "cors", 
        "Sec-Fetch-Site": "cross-site", 
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36", 
        "X-Amzn-Trace-Id": "Root=1-660fc23a-52234d884f1660a665985b24"
      }, 
      "origin": "126.28.198.5", 
      "url": "https://httpbin.org/get"
    }
    
  */  

</script>  
</body>
</html>